src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -1240,6 +1268,37 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
           }
       }
 
+    case (returnType, "flatMap", Array(Lambda(_, param, body))) =>
+      val localIdx = ec.a.length
+      val localA = ec.a
+      localA += null
+      val bodyFn = body.eval(ec.copy(st = ec.st + (param -> (localIdx, returnType))))
+
+      (returnType: @unchecked) match {
+        case TArray(_) =>
+          AST.evalCompose[IndexedSeq[_]](ec, lhs) { is =>
+            is.flatMap { elt =>
+              localA(localIdx) = elt
+              bodyFn().asInstanceOf[TraversableOnce[_]]
+            }
+          }
+        case TSet(_) =>
+          AST.evalCompose[Set[_]](ec, lhs) { s =>
+            s.flatMap { elt =>
+              localA(localIdx) = elt
+              bodyFn().asInstanceOf[TraversableOnce[_]]
+            }
+          }
+      }
+
+    case (returnType, "flatten", Array()) =>
+      (returnType: @unchecked) match {
+        case TArray(_) =>
+          AST.evalCompose[IndexedSeq[_]](ec, lhs) { is => is.asInstanceOf[IndexedSeq[IndexedSeq[_]]].flatten }
```

Comment:
I've fixed and added tests. Rather than calling to Scala's flatMap and flatten, I've implemented these functions in a way that checks for null collections and, if found, sets the entire result to null. Let me know your thoughts on the code replication. For example, I could use an iterator on the array as well, and put the def elsewhere (to Utils?), so that arrays and sets are handled by the same function in each case.